src/routing/tests/mod.rs

```diff
@@ -1437,6 +1443,119 @@ async fn message_to_self(dst: MessageDst) -> Result<()> {
+        .take(ELDER_SIZE - 1)
+        .collect();
+    let adult_peer = create_peer().with_age(MIN_AGE + 1);
+    let promoted_peer = create_peer().with_age(MIN_AGE + 3);
+
+    let sk_set0 = SecretKeySet::random();
+    let pk0 = sk_set0.secret_key().public_key();
+
+    let elders_info0 = EldersInfo::new(
+        iter::once(node.peer()).chain(other_elder_peers.clone()),
+        Prefix::default(),
+    );
+
+    let (mut section0, section_key_share) = create_section(&sk_set0, &elders_info0)?;
+
+    for peer in &[adult_peer, promoted_peer] {
+        let member_info = MemberInfo::joined(*peer);
+        let member_info = proven(sk_set0.secret_key(), member_info)?;
+        assert!(section0.update_member(member_info));
+    }
+
+    let demoted_peer = other_elder_peers.remove(0);
+
+    // Create `HandleConsensus` command for an `OurElders` vote. This will demote one of the
+    // current elders and promote the oldest peer.
+    let elders_info1 = EldersInfo::new(
+        iter::once(node.peer())
+            .chain(other_elder_peers.clone())
+            .chain(iter::once(promoted_peer)),
+        Prefix::default(),
+    );
+    let elder_names1: BTreeSet<_> = elders_info1.elders.keys().copied().collect();
+
+    let sk_set1 = SecretKeySet::random();
+    let pk1 = sk_set1.secret_key().public_key();
+
+    let proven_elders_info1 = proven(sk_set1.secret_key(), elders_info1)?;
+    let vote = Vote::OurElders(proven_elders_info1);
+    let signature = sk_set0
+        .secret_key()
+        .sign(&bincode::serialize(&vote.as_signable())?);
+    let proof = Proof {
+        signature,
+        public_key: pk0,
+    };
+
+    let (event_tx, mut event_rx) = mpsc::unbounded_channel();
+    let state = Approved::new(node, section0.clone(), Some(section_key_share), event_tx);
+    let stage = Stage::new(state, create_comm()?);
+
+    let commands = stage
+        .handle_command(Command::HandleConsensus { vote, proof })
+        .await?;
+
+    let mut sync_actual_recipients = HashSet::new();
+
+    for command in commands {
+        let (recipients, message) = match command {
+            Command::SendMessage {
+                recipients,
+                message,
+                ..
+            } => (recipients, message),
+            _ => continue,
+        };
+
+        let message = Message::from_bytes(&message)?;
+        let section = match message.variant() {
+            Variant::Sync { section, .. } => section,
+            _ => continue,
+        };
+
+        assert_eq!(section.chain().last_key(), &pk1);
+
+        // The message is trusted even by peers who don't yet know the new section key.
+        assert_matches!(
+            message.verify(iter::once((&Prefix::default(), &pk0))),
+            Ok(VerifyStatus::Full)
+        );
+
+        // Merging the section contained in the message with the original section succeeds.
+        assert_matches!(section0.clone().merge(section.clone()), Ok(()));
```

Comment:
do we also need to check the `merged section` now contains the latest key (pk1) ?

Reply:
Maybe, but we already checked that the `Section` in the message contains the latest key and I think it's OK to assume here the `merge` function works correctly because this is an unit test, not an integration test. The `merge` function should have its own unit test (although it currently doesn't :grimacing: )

Reply:
Yeah, I agree the merge function shall be covered within its own unit test.
we can add it in a separate PR later on.